core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -330,6 +330,10 @@ public static boolean containsNullable(RelDataType type) {
   public static RelDataType keepSourceTypeAndTargetNullability(RelDataType sourceRelDataType,
                                              RelDataType targetRelDataType,
                                              RelDataTypeFactory typeFactory) {
+    checkArgument(
+        (targetRelDataType.isStruct() && sourceRelDataType.isStruct())
+            || (!targetRelDataType.isStruct() && !sourceRelDataType.isStruct()),
+        "Input sourceRelDataType and targetRelDataType must have same struct type");
```

Comment:
I think that a better message is "one is a struct, while the other one is not".
"same struct type" implies that the type must be exactly the same.

Reply:
It has been modified as recommended.